fastlane/Fastfile

```diff
@@ -1101,3 +1106,20 @@ def trigger_buildkite_release_build(branch:, beta:)
     pipeline_file: 'release-builds.yml'
   )
 end
+
+def add_buildkite_analytics_token(xctestrun_path:)
+  require 'plist'
+
+  token_key = 'BUILDKITE_ANALYTICS_TOKEN'
+  token = ENV[token_key]
+  return if token.nil?
+
+  xctestrun = Plist.parse_xml(xctestrun_path)
+  xctestrun['TestConfigurations'].each do |configuration|
+    configuration['TestTargets'].each do |target|
+      target['EnvironmentVariables'][token_key] = token
+    end
+  end
+
+  File.write(xctestrun_path, Plist::Emit.dump(xctestrun))
+end
```

Comment:
Keen on feedback on how to improve this.

I'd like to merge the PR ASAP, so that we can start gather test analytics. I'll implement improvements in a followup.

---

I would have liked to keep the Buildkite-specific code out of the `Fastfile` to keep the automation there CI platform agnostic. But, we already have Buildkite-specific code in it, and it's convenient to manipulate the `xctestrun` file here, where we already have its path.

One improvement that comes to mind would be failing the lane if the token is not found. I'd want to refine the check, though, e.g. fail if on CI and token not found.

Reply:
What I don't like about this is that if we run this lane locally, this will write the token into the `xctestrun` file… which can then easily be commited into git accidentally.

I wonder if there isn't a possibility instead to pass parameters or env var when we run the test plan (via some flags in the `xcodebuild` invocation)? That way the token would only be passed on invocation and never written on disk, especially not on a file that is tracked by git.

I know that the `test_without_building` lane is supposed to only be run on CI, but we all know that it's not enough of an assumption to avoid the risk, especially as us Mobile Infra Engineer can often run those lanes on our local machines while debugging them…

Reply:
> I wonder if there isn't a possibility instead to pass parameters or env var when we run the test plan (via some flags in the xcodebuild invocation? Maybe xcargs:?). That way the token would only be passed on invocation and never written on disk, especially not on a file that is tracked by git.

I tried a number of `xcodebuild` invocation options [with this playground project](https://github.com/Automattic/TestPress-Apple) and trying different suggestions from [this StackOverflow questions](https://stackoverflow.com/questions/23284829/accessing-user-defined-variables-passed-in-from-xcodebuild-command-line/47762549#47762549). The `xctesrun` editing was the only one that worked for me.

> What I don't like about this is that if we run this lane locally, this will write the token into the `xctestrun` file… which can then easily be commited into git accidentally.

🤔 Does the fact that `xctestrun` file lives in the `DerivedData/` folder which is [ignored](https://github.com/wordpress-mobile/WordPress-iOS/blob/3d604c46e5899685d2060d4f3096d60ddfe12d09/.gitignore#L23), changes this concern?

*Lines configuring the `DerivedData` folder for `build-for-testing`*: https://github.com/woocommerce/woocommerce-ios/blob/2bab8e0b5908b246e6f4e6a52c4c39088bdd60ba/fastlane/Fastfile#L557-L564

*Lines uploading and downloading the `DerivedData` folder in CI*: https://github.com/woocommerce/woocommerce-ios/blob/2bab8e0b5908b246e6f4e6a52c4c39088bdd60ba/.buildkite/commands/build-for-testing.sh#L22, https://github.com/woocommerce/woocommerce-ios/blob/2bab8e0b5908b246e6f4e6a52c4c39088bdd60ba/.buildkite/commands/run-unit-tests.sh#L4

*Documentation for the `xctestrun` default location*—I don't even know if there's a way to change it:

> During the build-for-testing action, xcodebuild extracts parameters from the scheme provided to the build-for-testing action and writes the parameters to an xctestrun file in the built products directory
>
> – [source](https://keith.github.io/xcode-man-pages/xcodebuild.xctestrun.5.html)

Reply:
Ohhh right, if it's a file in DerivedData then it should be OK then.

I think I confused it with the `.xctestplan` file tbh.

The token would still be stored in a non-ephemeral way because the DerivedData are uploaded to s3 to pass them between build-for-testing and run-unit-tests, but this is way less concerning because the access to S3 is protected and not publicly accessible, so that shouldn't be a problem.